### ISSUE TITLE
The chrome can't get response after the app push or restart

### ISSUE THREAD

**Issue by huangered** [cloudfoundry/gorouter#112]
I meet a random strange issue about the sprint-boot and cloudfoundry.
1. I have two spring-boot projects, 1st is a angularjs website, 2nd is a rest api service.
the angularjs website will call the rest api in the second site.
2. two sites are deployed in the cloudfoundry.
3. The problem is when the service restart or redeploy, if I use the chrome browser to visit the site in the first, the angularjs site will return 404 when js call the rest service in cross domain.(and in the rest api cloundfoundry log, there is no request arrive)

The chrome response looks like:

**X-Cf-Routererror:unknown_route   (no this in firefox and curl)** i see this error is handled in the gorouter

General
Remote Address:10.62.90.144:80
Request URL:http://qaiaas-invenroty.10.62.90.144.xip.io/
Request Method:GET
Status Code:404 Not Found
Response Headers
view source
Content-Length:88
Content-Type:text/plain; charset=utf-8
Date:Fri, 08 Jan 2016 02:21:04 GMT
X-Cf-Requestid:283b81ce-2306-44fe-5a3d-2d71724f51e3
**X-Cf-Routererror:unknown_route   (no this in firefox and curl)**
Request Headers
view source
Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-US,en;q=0.8
Cache-Control:no-cache
Connection:keep-alive
Host:qaiaas-invenroty.10.62.90.144.xip.io
Pragma:no-cache

4. but if I use firefox browser, every thing is ok
5. then if I use chrome again, every thing is two.
**each instance is 2, and we have 3 vms in the cloudfoundry**
Did any people meet it before? and how to solve it?

**Comment by linlynn** [comment]
experts, if you can provide us some hints on this error, it will be much appreciated

**Comment by markstgodard** [comment]
Hi @huangered 

I do not recall ever seeing an issue like this, however let me clarify your scenario:

- You have two spring boot apps deployed to CF and running
- The first app is a web app that uses javascript to call a REST API running on the 2nd spring boot app
- The 2nd app (REST API) allows for cross domain calls

You are able to make requests to the first app (which calls the 2nd app via javascript / REST API call)

However if you restart or redeploy the 2nd app, then the requests from the first app's javascript / REST API receive an HTTP status code 404 and ONLY when using the Google Chrome browser ?

If this works in Firefox but not Chrome, then I suspect something is going on with the javascript running in the browser.

Which version of Cloud Foundry are you using?

Are you able to check the gorouter's access.log when you get a 404 in Chrome but not in Firefox?

If this a cloud foundry deployment which you are running ?
If so, you can dump the gorouter routing table to see which routes it is aware of:

i.e. curl http://<user>:<password>@<gorouter job / vm ip>:8080/routes 

If it works in Firefox but not in Chrome, then I suspect its something to do with the client side browser / javascript request.    Are you also saying that after the 2nd app redeploy,  Firefox works and also you can curl the rest API ?  but Chrome doesnt work?
Are you able to use Chrome's debugger to Inspect and see the Console output?

Basically if a request is received by the gorouter and it can't find the route in its routing table, it returns a 404.  


Your last comment on "then if I use chrome again, every thing is two. each instance is 2, and we have 3 vms in the cloudfoundry",  I dont quite understand, can you clarify ?

Each instance is 2 ?  Do you mean you now have 2 instances of the 2nd app ?
3 VMs in cloud foundry?  Are you referring to the gorouter jobs / vms ?

Thanks

**Comment by huangered** [comment]
Hi, @markstgodard 

You have two spring boot apps deployed to CF and running
The first app is a web app that uses javascript to call a REST API running on the 2nd spring boot app
The 2nd app (REST API) allows for cross domain calls
You are able to make requests to the first app (which calls the 2nd app via javascript / REST API call)
Peter: yes

However if you restart or redeploy the 2nd app, then the requests from the first app's javascript / REST API receive an HTTP status code 404 and ONLY when using the Google Chrome browser ?
Peter: yes

If this works in Firefox but not Chrome, then I suspect something is going on with the javascript running in the browser.

Which version of Cloud Foundry are you using?
Peter: 225

Are you able to check the gorouter's access.log when you get a 404 in Chrome but not in Firefox?
Peter: in the gorouter's log, there is no access when i use chrome

If it works in Firefox but not in Chrome, then I suspect its something to do with the client side browser / javascript request. Are you also saying that after the 2nd app redeploy, Firefox works and also you can curl the rest API ? but Chrome doesnt work?
Are you able to use Chrome's debugger to Inspect and see the Console output?
Peter: yes, curl and firefox works, in the chrome's debugger, the output for this request is 
listed in the first comment. The main important I think is 
X-Cf-Routererror:unknown_route

Your last comment on "then if I use chrome again, every thing is fine. The cf config is  each instance is 2, and we have 3 vms in the cloudfoundry", I dont quite understand, can you clarify ?
Peter: 1. use chrome, request fail, 2 use  firefox , request ok, 3 use chrome , then ok.
we have 3 vms for app in the cf, and each deployed app has 2 instances.

Thanks

**Comment by markstgodard** [comment]
Hi @huangered 

Correct, if you are getting a 404 response with these headers in the response, then its definitely getting to gorouter
```
X-Cf-Requestid:283b81ce-2306-44fe-5a3d-2d71724f51e3
X-Cf-Routererror:unknown_route
```

This means the gorouter instance that received the request could not find an entry in its internal routing table.
The routing table lookup is based on the request: Host and RequestURI 

Gorouter will log this information in both access log and gorouter.log on the VM that is running the gorouter instance. Should be under /var/vcap/sys/log/gorouter

For example, if I make a request to an app that does not exist in my CF:

curl doesnotexist.bosh-lite.com


**access.log:**
```
doesnotexist.bosh-lite.com - [28/01/2016:15:36:45 +0000] "GET / HTTP/1.1" 404 0 0 "-" "curl/7.43.0" 10.244.0.21:41550 x_forwarded_for:"192.168.50.1" x_forwarded_proto:"http" vcap_request_id:- response_time:- app_id:-
```

**gorouter.log:**
```json
{"timestamp":"1453995405.666825771","source":"gorouter.stdout","message":"gorouter.stdout.Request Handler","log_level":1,"data":{"Host":"doesnotexist.bosh-lite.com","Path":"/","RemoteAddr":"10.244.0.21:41550","X-Forwarded-For":["192.168.50.1"],"X-Forwarded-Proto":["http"]}}
{"timestamp":"1453995405.667052031","source":"gorouter.stdout","message":"gorouter.stdout.status ","log_level":1,"data":{"body":"404 Not Found: Requested route ('doesnotexist.bosh-lite.com') does not exist."}}
```

I am surprised you are not getting any logs in gorouter when you get a 404 but response includes the X-Cf-RouterError and X-Cf-RequestId.

Are you running a single gorouter instance or multiple? If more than one, can you check the logs on all gorouter instances?

The only thing that would affect the gorouter lookup of the route is the 
requests: Host and RequestURI,  so if this request is indeed getting to the gorouter, then I wanted to see what the gorouter log entries look like.

Cheers

**Comment by markstgodard** [comment]
Also, in the access.log it will include the **User-Agent** so we can see what the request looks like between all 3 browser type requests (i.e. any differences between Chrome, Firefox, Safari,  or curl)

for example, if I make 3 requests using 3 different browsers and if a I get 404:

access.log and User-Agent
**Chrome:**
doesnotexist.bosh-lite.com - [28/01/2016:15:51:43 +0000] "GET / HTTP/1.1" 404 0 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) **Chrome/47.0.2526.111** Safari/537.36" 10.244.0.21:49898 x_forwarded_for:"192.168.50.1" x_forwarded_proto:"http" vcap_request_id:- response_time:- app_id:-

**Safari**
doesnotexist.bosh-lite.com - [28/01/2016:15:52:04 +0000] "GET / HTTP/1.1" 404 0 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/601.2.7 (KHTML, like Gecko) Version/9.0.1 **Safari/601.2.7**" 10.244.0.21:50096 x_forwarded_for:"192.168.50.1" x_forwarded_proto:"http" vcap_request_id:- response_time:- app_id:-

**Firefox**
doesnotexist.bosh-lite.com - [28/01/2016:15:52:39 +0000] "GET / HTTP/1.1" 404 0 0 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:38.0) Gecko/20100101 **Firefox/38.0**" 10.244.0.21:50430 x_forwarded_for:"192.168.50.1" x_forwarded_proto:"http" vcap_request_id:- response_time:- app_id:-


What I would like to determine is,  when using Chrome, does the request get to gorouter (which it looks like it is) and what is shown in the access.log and gorouter.log for the request / 404 response.

From there we can look at the Host, Path, etc. and see if anything is malformed.

Cheers

**Comment by linlynn** [comment]
@markstgodard  we will provide necessary debug information once we get chance to re-produce it

**Comment by shalako** [comment]
Hello @huangered @linlynn we'll close this issue for now. Please reopen if you'd like to follow up.
